File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:

    int func(vector<int>weights,int capacity)
    {
        int days=1,total=0;
        for(int i=0;i<weights.size();i++)
        {
            if(total+weights[i]>capacity)
            {
                days=days+1;
                total=weights[i];
            }
            else
            {
                total+=weights[i];
            }
        }
        return days;
    }

    
    
    int shipWithinDays(vector<int>& weights, int days) {
        int low=*max_element(weights.begin(),weights.end());
        int high=accumulate(weights.begin(),weights.end(),0);
        while(low<=high)
        {
            int mid=low+(high-low)/2;
            int reqdays=func(weights,mid);
            if(reqdays<=days)
            {
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return low;
        
    }
};

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights: list[int], days: int) -> int:
        def fun(w):
            day=1
            sum=0
            for i in  weights:
                if sum+i<=w:
                    sum=sum+i

                else:
                    day=day+1
                    sum=i

            return day<=days


        l=max(weights)
        r=sum(weights)
        ans=0

        while l<=r:
            
            mid=(l+r)//2
                

            if(fun(mid)):
                ans=mid
                r=mid-1

            else:
                l=mid+1
            

        return ans


        

        